rust/hyperlane-base/src/contract_sync/cursors/mod.rs

```diff
@@ -242,27 +261,36 @@ where
         } else {
             Duration::from_secs(0)
         };
+        Ok(eta)
+    }
+}
+
+#[async_trait]
+impl<T> ContractSyncCursor<T> for RateLimitedContractSyncCursor<T>
+where
+    T: Send + Debug + 'static,
+{
+    async fn next_action(&mut self) -> Result<(CursorAction, Duration)> {
+        let eta = self.sync_eta()?;
 
         let rate_limit = self.get_rate_limit().await?;
```

Comment:
noting that in this call we update the `tip` and `last_tip_update` regardless of us ever successfully updating the cursor. so if there's an error and we're close to the tip we'll rate limit ourselves and only retry 30s from now

I think this is fine for now but as you continue to refactor we should probably consider this case

Reply:
I think this was the case before right? Less critical than other bits but I took a note